target_chains/starknet/contracts/src/reader.cairo

```diff
@@ -0,0 +1,333 @@
+        Result::Ok(value)
+    }
+
+    // TODO: skip without calculating values
+    fn skip(ref self: Reader, mut num_bytes: u8) -> Result<(), felt252> {
+        let mut result = Result::Ok(());
+        while num_bytes > 0 {
+            if num_bytes > 16 {
+                match self.read(16) {
+                    Result::Ok(_) => {},
+                    Result::Err(err) => {
+                        result = Result::Err(err);
+                        break;
+                    }
+                }
+                num_bytes -= 16;
+            } else {
+                match self.read(num_bytes) {
+                    Result::Ok(_) => {},
+                    Result::Err(err) => {
+                        result = Result::Err(err);
+                        break;
+                    }
+                }
+                break;
+            }
+        };
+        result
+    }
+
+    /// Reads the specified number of bytes as a new byte array.
+    fn read_bytes(ref self: Reader, num_bytes: usize) -> Result<ByteArray, felt252> {
+        let mut array: Array<bytes31> = array![];
+        let mut num_last_bytes = Option::None;
+        let mut num_remaining_bytes = num_bytes;
+        loop {
+            let r = self.read_bytes_iteration(num_remaining_bytes, ref array);
+            match r {
+                Result::Ok((
+                    num_read, eof
+                )) => {
+                    num_remaining_bytes -= num_read;
+                    if eof {
+                        num_last_bytes = Option::Some(Result::Ok(num_read));
+                        break;
+                    }
+                },
+                Result::Err(err) => {
+                    num_last_bytes = Option::Some(Result::Err(err));
+                    break;
+                }
+            }
+        };
+        // `num_last_bytes` is always set to Some before break.
+        let num_last_bytes = num_last_bytes.unwrap()?;
+        // num_last_bytes < 31
+        let num_last_bytes = num_last_bytes.try_into().expect(UNEXPECTED_OVERFLOW);
+        let array = ByteArrayImpl::new(array, num_last_bytes);
+        Result::Ok(array)
+    }
+
+    /// Returns number of remaining bytes to read.
+    fn len(ref self: Reader) -> usize {
+        let num_next_bytes = if self.next.is_some() {
+            16
+        } else {
+            0
+        };
+        self.num_current_bytes.into() + num_next_bytes + self.array.len()
+    }
+
+    /// Reads the specified number of bytes (up to 16) as a little endian unsigned integer.
+    fn read_le(ref self: Reader, num_bytes: u8) -> Result<u128, felt252> {
+        if num_bytes == 0 {
+            return Result::Ok(0);
+        }
+        let value = u128_byte_reverse(self.read(num_bytes)?)
+            / one_shift_left_bytes_u128(16 - num_bytes);
+        Result::Ok(value)
+    }
+
+    /// Reads and hashes all the remaining data.
+    fn keccak256(ref self: Reader) -> Result<u256, felt252> {
```

Comment:
I have a feeling that keccak doesn't belong here. Here can be just a function `read_bytes_as_le_array` or something and then the consumer should run keccak itself.

Reply:
I agree. I wrote a more flexible hashing implementation that will also be used for Pyth's Merkle tree proofs. Now `Reader` and `Hasher` are cleanly separated, and all byte order changes are now encapsulated in `Hasher`. This also simplified the Wormhole hashing code quite a bit.